File: src/main/java/org/example/graph/CourseScheduleII.java
package org.example.graph;

import java.util.*;

public class CourseScheduleII {

    Map<Integer, List<Integer>> prereq = new HashMap<>();
    Set<Integer> visit = new HashSet<>();
    Set<Integer> cycle = new HashSet<>();
    List<Integer> res = new ArrayList<>();
    public int[] findOrder(int numCourses, int[][] prerequisites) {
         for (int i=0; i < numCourses; i++)
             prereq.put(i, new ArrayList<>());
         for (int[] pre: prerequisites)
             prereq.get(pre[0]).add(pre[1]);

         for (int crs=0; crs < numCourses; crs++) {
             if (!dfs(crs)) return new int[0];
         }

         return res.stream().mapToInt(i -> i).toArray();
    }

    private boolean dfs(int crs) {
        if (cycle.contains(crs)) return false;
        if (visit.contains(crs)) return true;

        cycle.add(crs);
        for (int pre: prereq.get(crs))
            if (!dfs(pre)) return false;
        cycle.remove(crs);
        visit.add(crs);
        res.add(crs);
        return true;
    }
}

File: src/main/java/org/example/graph/CourseScheduleII.py
from typing import List


class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        prereq = {c: [] for c in range(numCourses)}
        for crs, pre in prerequisites:
            prereq[crs].append(pre)

        # a course has 3 possible states:
        # visited: crs has been added to output
        # visiting: crs not added to output, but added to cycle
        # unvisited: crs not added to output or cycle
        output = []
        visit, cycle = set(), set()

        def dfs(crs):
            if crs in cycle: return False
            if crs in visit: return True

            cycle.add(crs)
            for pre in prereq[crs]:
                if not dfs(pre): return False
            cycle.remove(crs)
            visit.add(crs)
            output.append(crs)
            return True

        for c in range(numCourses):
            if not dfs(c): return []
        return output
